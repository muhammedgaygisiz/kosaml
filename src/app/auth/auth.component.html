<mat-card class="auth-card">
    <div style="text-align: center;" *ngIf="(isLoading$ | async)">
        <kosaml-loading-spinner></kosaml-loading-spinner>
    </div>
    <form [formGroup]="authForm" (ngSubmit)="onSubmit()" *ngIf="!(isLoading$ | async)">
        <div class="formFieldContainer">
            <mat-form-field>
                <input matInput placeholder="Email" name="email" formControlName="email" [errorStateMatcher]="matcher">
                <mat-error *ngIf="emailFormControl.hasError('email') && !emailFormControl.hasError('required')">
                    Please enter a valid email address
                </mat-error>
                <mat-error *ngIf="emailFormControl.hasError('required')">
                    Email is <strong>required</strong>
                </mat-error>
                <mat-error *ngIf="(authError$ | async) === true">
                    Email or password is not correct
                </mat-error>
            </mat-form-field>
            <mat-form-field>
                <input matInput placeholder="Password" type="password" name="password" formControlName="password">
                <mat-error
                    *ngIf="passwordFormControl.hasError('minlength') && !passwordFormControl.hasError('required')">
                    Passord must be at least 6 characters long
                </mat-error>
                <mat-error *ngIf="passwordFormControl.hasError('required')">
                    Passoword is <strong>required</strong>
                </mat-error>
                <mat-error *ngIf="(authError$ | async) === true">
                    Email or password is not correct
                </mat-error>
            </mat-form-field>
            <div class="buttonContainer">
                <button mat-raised-button color="primary" type="submit"
                    [disabled]="!authForm.valid">{{ isLoginMode ? 'Login' : 'Sign Up' }}</button>
                <button mat-button color="primary" (click)="onSwitchMode()" type="button">Switch to
                    {{ isLoginMode ? 'Sign Up' : 'Login' }}</button>
            </div>
            <mat-error *ngIf="(isAuthError$ | async) === true">
                Authentication failed
            </mat-error>
        </div>
    </form>
</mat-card>