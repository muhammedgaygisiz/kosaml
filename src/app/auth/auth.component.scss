.auth-card {
    margin: 30px auto;
    max-width: 400px;
    min-width: 140px;
}

.formFieldContainer {
    display: flex;
    flex-direction: column;
}

.buttonContainer {
    margin-top: 12px;
}

.buttonContainer button {
    margin-right: 8px;
}