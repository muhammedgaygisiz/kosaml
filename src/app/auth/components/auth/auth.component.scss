.formFieldContainer {
    display: flex;
    flex-direction: column;
}

.buttonContainer {
    margin: 12px 0;
}

.buttonContainer button {
    margin-right: 8px;
}