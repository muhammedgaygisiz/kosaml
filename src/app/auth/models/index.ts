import { Credentials } from './credentials.model';
import { User } from './user.model';

export { Credentials, User };

