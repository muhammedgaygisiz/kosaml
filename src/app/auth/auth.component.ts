import { Component, OnInit } from '@angular/core';
import { FormControl, FormGroup, NgForm, Validators } from '@angular/forms';
import { Store } from '@ngrx/store';
import * as AuthActions from '../auth/store/auth.actions';
import * as fromApp from '../store/app.reducer';



@Component({
  selector: 'kosaml-auth',
  templateUrl: './auth.component.html',
  styleUrls: ['./auth.component.scss']
})
export class AuthComponent implements OnInit {
  isLoginMode = true;
  authForm: FormGroup;

  constructor(
    private store: Store<fromApp.AppState>,
  ) { }

  ngOnInit() {
    this.authForm = new FormGroup({
      email: new FormControl(null, [Validators.required, Validators.email]),
      password: new FormControl(null, [Validators.required, Validators.minLength(6)]),
    });
  }

  onSwitchMode() {
    this.isLoginMode = !this.isLoginMode;
  }

  onSubmit(form: NgForm) {
    if (!form.value) {
      return;
    }

    const email = form.value.email;
    const password = form.value.password;

    if (this.isLoginMode) {
      this.store.dispatch(new AuthActions.LoginStart(
        { email, password }
      ));
    } else {
      this.store.dispatch(new AuthActions.SignupStart(
        { email, password }
      ));
    }

    form.reset();
  }



}
