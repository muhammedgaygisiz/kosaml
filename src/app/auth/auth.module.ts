import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatCardModule } from '@angular/material/card';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { SharedModule } from '../shared/shared.module';
import { AuthComponent } from './components';
import { AuthPageComponent } from './containers/auth-page.component';
import { AuthEffects } from './effects/auth.effects';
import * as fromAuth from './reducers/auth.reducer';

export const COMPONENTS = [
    AuthPageComponent,
    AuthComponent,
];

@NgModule({
    declarations: COMPONENTS,
    imports: [
        SharedModule,
        ReactiveFormsModule,
        MatFormFieldModule,
        MatCardModule,
        MatInputModule,
        MatButtonModule,
        HttpClientModule,
        RouterModule.forChild([{ path: '', component: AuthPageComponent }]),
        StoreModule.forFeature(fromAuth.authFeatureKey, fromAuth.authReducer),
        EffectsModule.forFeature([AuthEffects])
    ]
})
export class AuthModule {

}
