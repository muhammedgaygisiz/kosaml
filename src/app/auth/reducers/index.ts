import * as fromAuth from './auth.reducer';

export { fromAuth };
export default fromAuth;
