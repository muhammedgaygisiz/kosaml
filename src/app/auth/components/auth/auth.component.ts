import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { Credentials } from '../../models';
import { AuthErrorMatcher } from './AuthErrorMatcher';

@Component({
  selector: 'kosaml-auth',
  templateUrl: './auth.component.html',
  styleUrls: ['./auth.component.scss'],
})
export class AuthComponent implements OnInit {

  @Input()
  isLoginMode = true;

  authForm: FormGroup;

  emailFormControl: FormControl = new FormControl(null, [Validators.required, Validators.email]);
  passwordFormControl: FormControl = new FormControl(null, [
    Validators.required,
    Validators.minLength(6),
  ]);
  matcher = new AuthErrorMatcher();

  @Input()
  isAuthError: boolean;

  @Output()
  loginSubmitted = new EventEmitter<Credentials>();

  @Output()
  registrationSubmitted = new EventEmitter<Credentials>();

  constructor() { }

  ngOnInit() {
    this.authForm = new FormGroup({
      email: this.emailFormControl,
      password: this.passwordFormControl,
    });
  }

  onSwitchMode() {
    this.isLoginMode = !this.isLoginMode;
  }

  onSubmit() {
    if (!this.authForm.value) {
      return;
    }

    const email = this.authForm.value.email;
    const password = this.authForm.value.password;

    if (this.isLoginMode) {
      this.loginSubmitted.emit({ email, password });
    } else {
      this.registrationSubmitted.emit({ email, password });
    }

    this.authForm.reset();
  }

  get controls() {
    return this.authForm.controls;
  }
}
