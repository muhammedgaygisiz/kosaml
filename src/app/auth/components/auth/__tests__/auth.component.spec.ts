import { DebugElement, NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { SharedModule } from 'src/app/shared';
import { AuthComponent } from '../auth.component';


const getDebugElementForFieldName =
    (
        fieldName: string,
        fixture: ComponentFixture<AuthComponent>,
    ): DebugElement =>
        fixture.debugElement.query(By.css(`input[name='${fieldName}']`));

/**
 * Create custom DOM event the old fashioned way
 *
 * https://developer.mozilla.org/en-US/docs/Web/API/Event/initEvent
 * Although officially deprecated, some browsers (phantom) don't accept the preferred "new Event(eventName)"
 */
const newEvent = (eventName: string, bubbles = false, cancelable = false) => {
    const evt = document.createEvent('CustomEvent');  // MUST be 'CustomEvent'
    evt.initCustomEvent(eventName, bubbles, cancelable, null);
    return evt;
}


describe('PageComponent', () => {
    let fixture: ComponentFixture<AuthComponent>;
    let instance: AuthComponent;

    beforeEach(async () => {
        TestBed.configureTestingModule({
            declarations: [AuthComponent],
            imports: [SharedModule],
            schemas: [NO_ERRORS_SCHEMA]
        })
            .compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(AuthComponent);

        instance = fixture.componentInstance;
    })

    it('should create', () => {
        fixture.detectChanges();

        expect(instance).toBeTruthy();
    });

    it('should render with login button disabled', () => {
        fixture.detectChanges();

        expect(fixture).toMatchSnapshot();
    });

    it('should enable button when filled out', () => {
        fixture.detectChanges();

        const emailField = getDebugElementForFieldName('email', fixture);
        const pwdField = getDebugElementForFieldName('password', fixture);

        (emailField.nativeElement as HTMLInputElement).value = 'l@l';
        emailField.triggerEventHandler('change', {});

        // emailField.dispatchEvent(newEvent('input'))

        (pwdField.nativeElement as HTMLInputElement).value = 'Test';
        pwdField.triggerEventHandler('change', {});

        // pwdField.dispatchEvent(newEvent('input'))

        fixture.detectChanges();

        fixture.whenStable().then(() => {
            expect(fixture).toMatchSnapshot();
        })
    })
});
