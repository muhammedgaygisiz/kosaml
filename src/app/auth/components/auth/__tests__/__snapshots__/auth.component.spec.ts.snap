// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageComponent should enable button when filled out 1`] = `
<kosaml-auth
  authForm={[Function FormGroup]}
  emailFormControl={[Function FormControl]}
  isLoginMode={[Function Boolean]}
  loginSubmitted={[Function EventEmitter]}
  matcher={[Function AuthErrorMatcher]}
  passwordFormControl={[Function FormControl]}
  registrationSubmitted={[Function EventEmitter]}
>
  <kosaml-card>
    <mat-card
      class="card mat-card"
    >
      <form
        class="formFieldContainer ng-untouched ng-pristine ng-invalid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <mat-form-field
          class="mat-form-field ng-tns-c6-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid mat-form-field-should-float"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-invalid"
                  formcontrolname="email"
                  id="mat-input-4"
                  matinput=""
                  name="email"
                  ng-reflect-error-state-matcher="[object Object]"
                  ng-reflect-name="email"
                  ng-reflect-placeholder="Email"
                  placeholder="Email"
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-4"
                    class="mat-form-field-label ng-tns-c6-0 ng-star-inserted"
                    for="mat-input-4"
                    id="mat-form-field-label-9"
                    ng-reflect-disabled="true"
                    ng-reflect-ng-switch="false"
                  >
                    
                    
                    <span
                      class="ng-tns-c6-0 ng-star-inserted"
                    >
                      Email
                    </span>
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-underline ng-tns-c6-0 ng-star-inserted"
            >
              <span
                class="mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-form-field-subscript-wrapper"
              ng-reflect-ng-switch="hint"
            >
              
              
              <div
                class="mat-form-field-hint-wrapper ng-tns-c6-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
              >
                
                <div
                  class="mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-form-field>
        <mat-form-field
          class="mat-form-field ng-tns-c6-1 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid mat-form-field-should-float"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-invalid"
                  formcontrolname="password"
                  id="mat-input-5"
                  matinput=""
                  name="password"
                  ng-reflect-name="password"
                  ng-reflect-placeholder="Password"
                  ng-reflect-type="password"
                  placeholder="Password"
                  type="password"
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-5"
                    class="mat-form-field-label ng-tns-c6-1 ng-star-inserted"
                    for="mat-input-5"
                    id="mat-form-field-label-11"
                    ng-reflect-disabled="true"
                    ng-reflect-ng-switch="false"
                  >
                    
                    
                    <span
                      class="ng-tns-c6-1 ng-star-inserted"
                    >
                      Password
                    </span>
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-underline ng-tns-c6-1 ng-star-inserted"
            >
              <span
                class="mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-form-field-subscript-wrapper"
              ng-reflect-ng-switch="hint"
            >
              
              
              <div
                class="mat-form-field-hint-wrapper ng-tns-c6-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
              >
                
                <div
                  class="mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-form-field>
        <div
          class="buttonContainer"
        >
          <button
            class="mat-raised-button mat-button-base mat-primary"
            color="primary"
            disabled="true"
            mat-raised-button=""
            ng-reflect-color="primary"
            ng-reflect-disabled="true"
            type="submit"
          >
            <span
              class="mat-button-wrapper"
            >
               Login 
            </span>
            <div
              class="mat-button-ripple mat-ripple"
              matripple=""
              ng-reflect-centered="false"
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLButtonElement]"
            />
            <div
              class="mat-button-focus-overlay"
            />
          </button>
          <button
            class="mat-button mat-button-base mat-primary"
            color="primary"
            mat-button=""
            ng-reflect-color="primary"
            type="button"
          >
            <span
              class="mat-button-wrapper"
            >
               Switch to Sign Up 
            </span>
            <div
              class="mat-button-ripple mat-ripple"
              matripple=""
              ng-reflect-centered="false"
              ng-reflect-disabled="false"
              ng-reflect-trigger="[object HTMLButtonElement]"
            />
            <div
              class="mat-button-focus-overlay"
            />
          </button>
        </div>
        
      </form>
    </mat-card>
  </kosaml-card>
</kosaml-auth>
`;

exports[`PageComponent should render with login button disabled 1`] = `
<kosaml-auth
  authForm={[Function FormGroup]}
  emailFormControl={[Function FormControl]}
  isLoginMode={[Function Boolean]}
  loginSubmitted={[Function EventEmitter]}
  matcher={[Function AuthErrorMatcher]}
  passwordFormControl={[Function FormControl]}
  registrationSubmitted={[Function EventEmitter]}
>
  <kosaml-card>
    <mat-card
      class="card mat-card"
    >
      <form
        class="formFieldContainer ng-untouched ng-pristine ng-invalid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <mat-form-field
          class="mat-form-field ng-tns-c3-0 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-invalid"
                  formcontrolname="email"
                  id="mat-input-2"
                  matinput=""
                  name="email"
                  ng-reflect-error-state-matcher="[object Object]"
                  ng-reflect-name="email"
                  ng-reflect-placeholder="Email"
                  placeholder="Email"
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-2"
                    class="mat-form-field-label ng-tns-c3-0 mat-empty mat-form-field-empty ng-star-inserted"
                    for="mat-input-2"
                    id="mat-form-field-label-5"
                    ng-reflect-disabled="true"
                    ng-reflect-ng-switch="false"
                  >
                    
                    
                    <span
                      class="ng-tns-c3-0 ng-star-inserted"
                    >
                      Email
                    </span>
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-underline ng-tns-c3-0 ng-star-inserted"
            >
              <span
                class="mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-form-field-subscript-wrapper"
              ng-reflect-ng-switch="hint"
            >
              
              
              <div
                class="mat-form-field-hint-wrapper ng-tns-c3-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
              >
                
                <div
                  class="mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-form-field>
        <mat-form-field
          class="mat-form-field ng-tns-c3-1 mat-primary mat-form-field-type-mat-input mat-form-field-appearance-legacy mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-invalid"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              
              
              <div
                class="mat-form-field-infix"
              >
                <input
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-input-element mat-form-field-autofill-control cdk-text-field-autofill-monitored ng-untouched ng-pristine ng-invalid"
                  formcontrolname="password"
                  id="mat-input-3"
                  matinput=""
                  name="password"
                  ng-reflect-name="password"
                  ng-reflect-placeholder="Password"
                  ng-reflect-type="password"
                  placeholder="Password"
                  type="password"
                />
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                  <label
                    aria-owns="mat-input-3"
                    class="mat-form-field-label ng-tns-c3-1 mat-empty mat-form-field-empty ng-star-inserted"
                    for="mat-input-3"
                    id="mat-form-field-label-7"
                    ng-reflect-disabled="true"
                    ng-reflect-ng-switch="false"
                  >
                    
                    
                    <span
                      class="ng-tns-c3-1 ng-star-inserted"
                    >
                      Password
                    </span>
                    
                    
                  </label>
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-underline ng-tns-c3-1 ng-star-inserted"
            >
              <span
                class="mat-form-field-ripple"
              />
            </div>
            <div
              class="mat-form-field-subscript-wrapper"
              ng-reflect-ng-switch="hint"
            >
              
              
              <div
                class="mat-form-field-hint-wrapper ng-tns-c3-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
              >
                
                <div
                  class="mat-form-field-hint-spacer"
                />
              </div>
            </div>
          </div>
        </mat-form-field>
        <div
          class="buttonContainer"
        >
          <button
            class="mat-raised-button mat-button-base mat-primary"
            color="primary"
            disabled="true"
            mat-raised-button=""
            ng-reflect-color="primary"
            ng-reflect-disabled="true"
            type="submit"
          >
            <span
              class="mat-button-wrapper"
            >
               Login 
            </span>
            <div
              class="mat-button-ripple mat-ripple"
              matripple=""
              ng-reflect-centered="false"
              ng-reflect-disabled="true"
              ng-reflect-trigger="[object HTMLButtonElement]"
            />
            <div
              class="mat-button-focus-overlay"
            />
          </button>
          <button
            class="mat-button mat-button-base mat-primary"
            color="primary"
            mat-button=""
            ng-reflect-color="primary"
            type="button"
          >
            <span
              class="mat-button-wrapper"
            >
               Switch to Sign Up 
            </span>
            <div
              class="mat-button-ripple mat-ripple"
              matripple=""
              ng-reflect-centered="false"
              ng-reflect-disabled="false"
              ng-reflect-trigger="[object HTMLButtonElement]"
            />
            <div
              class="mat-button-focus-overlay"
            />
          </button>
        </div>
        
      </form>
    </mat-card>
  </kosaml-card>
</kosaml-auth>
`;
